common/persistence/visibilityDualManager.go

```diff
@@ -0,0 +1,282 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package persistence
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/dynamicconfig"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/types"
+)
+
+type (
+	visibilityDualManager struct {
+		logger              log.Logger
+		dbVisibilityManager VisibilityManager
+		esVisibilityManager VisibilityManager
+		readModeIsFromES    dynamicconfig.BoolPropertyFnWithDomainFilter
+		writeMode           dynamicconfig.StringPropertyFn
+	}
+)
+
+var _ VisibilityManager = (*visibilityDualManager)(nil)
+
+// NewVisibilityDualManager create a visibility manager that operate on DB or ElasticSearch based on dynamic config.
+func NewVisibilityDualManager(
+	dbVisibilityManager VisibilityManager, // one of the VisibilityManager can be nil
+	esVisibilityManager VisibilityManager, // one of the VisibilityManager can be nil
+	readModeIsFromES dynamicconfig.BoolPropertyFnWithDomainFilter,
+	visWritingMode dynamicconfig.StringPropertyFn,
+	logger log.Logger,
+) VisibilityManager {
+	if dbVisibilityManager == nil && esVisibilityManager == nil {
+		logger.Fatal("require one of dbVisibilityManager or esVisibilityManager")
+		return nil
+	}
+	return &visibilityDualManager{
+		dbVisibilityManager: dbVisibilityManager,
+		esVisibilityManager: esVisibilityManager,
+		readModeIsFromES:    readModeIsFromES,
+		writeMode:           visWritingMode,
+		logger:              logger,
+	}
+}
+
+func (v *visibilityDualManager) Close() {
+	if v.dbVisibilityManager != nil {
+		v.dbVisibilityManager.Close()
+	}
+	if v.esVisibilityManager != nil {
+		v.esVisibilityManager.Close()
+	}
+}
+
+func (v *visibilityDualManager) GetName() string {
+	if v.esVisibilityManager != nil {
+		return v.esVisibilityManager.GetName()
+	}
+	return v.dbVisibilityManager.GetName()
+}
+
+func (v *visibilityDualManager) RecordWorkflowExecutionStarted(
+	ctx context.Context,
+	request *RecordWorkflowExecutionStartedRequest,
+) error {
+	return v.chooseVisibilityManagerForWrite(
+		func() error {
+			return v.dbVisibilityManager.RecordWorkflowExecutionStarted(ctx, request)
+		},
+		func() error {
+			return v.esVisibilityManager.RecordWorkflowExecutionStarted(ctx, request)
+		},
+	)
+}
+
+func (v *visibilityDualManager) RecordWorkflowExecutionClosed(
+	ctx context.Context,
+	request *RecordWorkflowExecutionClosedRequest,
+) error {
+	return v.chooseVisibilityManagerForWrite(
+		func() error {
+			return v.dbVisibilityManager.RecordWorkflowExecutionClosed(ctx, request)
+		},
+		func() error {
+			return v.esVisibilityManager.RecordWorkflowExecutionClosed(ctx, request)
+		},
+	)
+}
+
+func (v *visibilityDualManager) DeleteWorkflowExecution(
+	ctx context.Context,
+	request *VisibilityDeleteWorkflowExecutionRequest,
+) error {
+	return v.chooseVisibilityManagerForWrite(
```

Comment:
Original behavior deletes workflows from all available storages, which is better for data cleaning. What do you think of it?

Reply:
Yes I changed that. I don't see it necessary. People uses dual write mode only for migration. After the migration, one of them will be removed anyway. Changing it here will make the behavior consistent with the configuration, or under the expectation of the configuration. It would be very hard to explain/documenting to have such a exceptional behavior in deleting.